crates/sp-consensus-subspace/src/verification.rs

```diff
@@ -34,29 +32,58 @@ use subspace_solving::{
     verify_local_challenge, verify_tag_signature, SubspaceCodec,
 };
 
+/// Errors encountered by the Subspace consensus primitives.
+#[derive(Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "thiserror", derive(thiserror::Error))]
+pub enum Error {
+    /// Tag verification failed
+    #[cfg_attr(feature = "thiserror", error("Invalid tag for salt"))]
+    InvalidTag,
+
+    /// Piece encoding is invalid
+    #[cfg_attr(feature = "thiserror", error("Invalid piece encoding"))]
+    InvalidPieceEncoding,
+
+    /// Piece verification failed
+    #[cfg_attr(feature = "thiserror", error("Invalid piece"))]
+    InvalidPiece,
+
+    /// Invalid Local challenge
+    #[cfg_attr(feature = "thiserror", error("Invalid local challenge"))]
+    InvalidLocalChallenge(SignatureError),
+
+    /// Solution is outside the challenge range
+    #[cfg_attr(feature = "thiserror", error("Solution is outside the solution range"))]
+    OutsideSolutionRange,
+
+    /// Invalid solution signature
+    #[cfg_attr(feature = "thiserror", error("Invalid solution signature"))]
+    InvalidSolutionSignature(SignatureError),
+
+    /// Solution is outside the MaxPlot
+    #[cfg_attr(feature = "thiserror", error("Solution is outside max plot"))]
+    OutsideMaxPlot,
+}
+
 /// Check the reward signature validity.
 pub fn check_reward_signature(
     hash: &[u8],
-    signature: &FarmerSignature,
-    public_key: &FarmerPublicKey,
+    signature: impl AsRef<[u8]>,
+    public_key: impl AsRef<[u8]>,
```

Comment:
We generally don't use `impl Trait` in the codebase.
We also already have generic public key and reward signature types in `subspace-core-primitives`, it would be better to use those at API level instead of taking random memory slices of unknown length.

Reply:
Fair to be consistent and not use `impl Trait` but I'm not convinced about using the wrapper public key and signature here  mainly because
- as all of schnorrkel types only require a slice reference 🤮  instead of array type and hence, a trait bound makes for more cleaner approach IMO since the errors are handled by the schnorrkel and api user would be handling those
-  `app::Public and app::Signature` wrappers do not seem to provide any array conversions to assist with conversions to wrapper publickey and signature. So it would be awkard to convert to slice and then Wrapper types and then to slice for schnorrkel types.
- I also see there is a TODO to make these app::* types more easier to convert to schnorrkel types but feels like that could be handled in a different PR altogether